File: hu/codes/java/chapter_stack_and_queue/deque.java
/**
 * File: deque.java
 * Created Time: 2022-11-25
 * Author: krahets (krahets@163.com)
 */

package chapter_stack_and_queue;

import java.util.*;

public class deque {
    public static void main(String[] args) {
        /* Kétirányú sor hosszának lekérése */
        Deque<Integer> deque = new LinkedList<>();
        deque.offerLast(3);
        deque.offerLast(2);
        deque.offerLast(5);
        System.out.println("Double-ended queue deque = " + deque);

        /* Elem frissítése */
        int peekFirst = deque.peekFirst();
        System.out.println("Front element peekFirst = " + peekFirst);
        int peekLast = deque.peekLast();
        System.out.println("Rear element peekLast = " + peekLast);

        /* Elemek sorba rakása */
        deque.offerLast(4);
        System.out.println("After element 4 enqueues at rear, deque = " + deque);
        deque.offerFirst(1);
        System.out.println("After element 1 enqueues at front, deque = " + deque);

        /* Elem kivétele */
        int popLast = deque.pollLast();
        System.out.println("Rear dequeue element = " + popLast + ", after rear dequeue, deque = " + deque);
        int popFirst = deque.pollFirst();
        System.out.println("Front dequeue element = " + popFirst + ", after front dequeue, deque = " + deque);

        /* Kétirányú sor hosszának lekérése */
        int size = deque.size();
        System.out.println("Double-ended queue length size = " + size);

        /* Kétirányú sor üresség ellenőrzése */
        boolean isEmpty = deque.isEmpty();
        System.out.println("Double-ended queue is empty = " + isEmpty);
    }
}

File: hu/codes/python/chapter_stack_and_queue/deque.py
"""
File: deque.py
Created Time: 2022-11-29
Author: Peng Chen (pengchzn@gmail.com)
"""

from collections import deque

"""Fő kód"""
if __name__ == "__main__":
    # Kétirányú sor inicializálása
    deq: deque[int] = deque()

    # Elemek sorba állítása
    deq.append(2)  # Hozzáadás a sor végéhez
    deq.append(5)
    deq.append(4)
    deq.appendleft(3)  # Hozzáadás a sor elejéhez
    deq.appendleft(1)
    print("kétirányú sor deque =", deq)

    # Elemek elérése
    front: int = deq[0]  # A sor elejének eleme
    print("Sor eleje front =", front)
    rear: int = deq[-1]  # A sor végének eleme
    print("Sor vége rear =", rear)

    # Elemek kivétele
    pop_front: int = deq.popleft()  # Az sor elejéről kivétel
    print("Sor elejéről kivett elem pop_front =", pop_front)
    print("Sor elejéről kivétel után deque =", deq)
    pop_rear: int = deq.pop()  # A sor végéről kivétel
    print("Sor végéről kivett elem pop_rear =", pop_rear)
    print("Sor végéről kivétel után deque =", deq)

    # A kétirányú sor hosszának lekérése
    size: int = len(deq)
    print("Kétirányú sor hossza size =", size)

    # Ellenőrzi, hogy a kétirányú sor üres-e
    is_empty: bool = len(deq) == 0
    print("A kétirányú sor üres-e =", is_empty)
